src/ripple/basics/impl/make_SSLContext.cpp

```diff
@@ -45,195 +82,88 @@ namespace detail {
-    {
-        EVP_PKEY_free(evp_pkey);
-    }
-};
-
-template <>
-struct custom_delete<X509>
-{
-    explicit custom_delete() = default;
-
-    void
-    operator()(X509* x509) const
-    {
-        X509_free(x509);
-    }
-};
-
-template <>
-struct custom_delete<DH>
-{
-    explicit custom_delete() = default;
-
-    void
-    operator()(DH* dh) const
-    {
-        DH_free(dh);
-    }
-};
-
-template <class T>
-using custom_delete_unique_ptr = std::unique_ptr<T, custom_delete<T>>;
-
-// RSA
-
-using rsa_ptr = custom_delete_unique_ptr<RSA>;
-
-static rsa_ptr
-rsa_generate_key(int n_bits)
+static void
+initAnonymous(boost::asio::ssl::context& context)
 {
-#if OPENSSL_VERSION_NUMBER >= 0x00908000L
-    BIGNUM* bn = BN_new();
-    BN_set_word(bn, RSA_F4);
-
-    RSA* rsa = RSA_new();
-    if (RSA_generate_key_ex(rsa, n_bits, bn, nullptr) != 1)
-    {
-        RSA_free(rsa);
-        rsa = nullptr;
-    }
+    using namespace openssl;
 
-    BN_free(bn);
-#else
-    RSA* rsa = RSA_generate_key(n_bits, RSA_F4, nullptr, nullptr);
-#endif
+    static auto defaultRSA = []() {
+        BIGNUM* bn = BN_new();
+        BN_set_word(bn, RSA_F4);
 
-    if (rsa == nullptr)
-        LogicError("RSA_generate_key failed");
+        auto rsa = RSA_new();
 
-    return rsa_ptr(rsa);
-}
+        if (!rsa)
+            LogicError("RSA_new failed");
 
-// EVP_PKEY
+        if (RSA_generate_key_ex(rsa, defaultRSAKeyBits, bn, nullptr) != 1)
+            LogicError("RSA_generate_key_ex failure");
 
-using evp_pkey_ptr = custom_delete_unique_ptr<EVP_PKEY>;
+        BN_free(bn);
 
-static evp_pkey_ptr
-evp_pkey_new()
-{
-    EVP_PKEY* evp_pkey = EVP_PKEY_new();
+        return rsa;
+    }();
 
-    if (evp_pkey == nullptr)
-        LogicError("EVP_PKEY_new failed");
+    static auto defaultPrivateKey = []() {
+        auto pkey = EVP_PKEY_new();
 
-    return evp_pkey_ptr(evp_pkey);
-}
+        if (!pkey)
+            LogicError("EVP_PKEY_new failed");
 
-static void
-evp_pkey_assign_rsa(EVP_PKEY* evp_pkey, rsa_ptr rsa)
-{
-    if (!EVP_PKEY_assign_RSA(evp_pkey, rsa.get()))
-        LogicError("EVP_PKEY_assign_RSA failed");
+        // We need to up the reference count of here, since we are retaining a
+        // copy of the key for (potential) reuse.
+        if (RSA_up_ref(defaultRSA) != 1)
+            LogicError(
+                "EVP_PKEY_assign_RSA: incrementing reference count failed");
 
-    rsa.release();
-}
+        if (!EVP_PKEY_assign_RSA(pkey, defaultRSA))
+            LogicError("EVP_PKEY_assign_RSA failed");
 
-// X509
+        return pkey;
+    }();
 
-using x509_ptr = custom_delete_unique_ptr<X509>;
+    static auto defaultCert = []() {
+        auto x509 = X509_new();
 
-static x509_ptr
-x509_new()
-{
-    X509* x509 = X509_new();
+        if (x509 == nullptr)
+            LogicError("X509_new failed");
 
-    if (x509 == nullptr)
-        LogicError("X509_new failed");
+        X509_set_version(x509, NID_X509);
```

Comment:
Still the wrong version:

https://github.com/ripple/rippled/issues/4007

Reply:
Thanks for reminding me of this. I'll fix this.